File: classes02/cpp/hueHistogram.cpp
#include <opencv2/opencv.hpp>
#include <iostream>

using namespace std;
using namespace cv;

int main(void)
{
    // 读取一张图像
    Mat img=imread("../data/images/capsicum.jpg");

    // 将图像转化为HSV颜色空间
    Mat hsvImage;
    cvtColor(img,hsvImage,COLOR_BGR2HSV);

    // 拆分图像的通道
    vector<Mat> channels(3);
    split(hsvImage, channels);

    imshow("Image", img);

    // 初始化参数
    int histSize=180;
    float range[]={0,179};
    const float *ranges[]={range};

    // 计算直方图
    MatND hist;
    calcHist(&channels[0],1,0,Mat(),hist,1,&histSize,ranges);
    cout<<hist<<endl;
    // 绘制直方图
    int hist_w=histSize*15; //直方图的宽度
    int hist_h=400;  //直方图的高度
    int bin_w=cvRound((double)hist_w/histSize);

    // 将直方图构建为一个图片
    Mat histImage(hist_h,hist_w,CV_8UC3,Scalar(255,255,255));
    normalize(hist, hist,0,histImage.rows,NORM_MINMAX,-1,Mat());

    // 绘制x轴
    line(histImage,Point(0,hist_h-35), Point(0,hist_h-25), Scalar(0,0,0), 2, 8, 0);
    putText(histImage, "0", Point(0,hist_h-5), FONT_HERSHEY_COMPLEX, 0.5, Scalar(0,0,0), 1, LINE_AA);

    for(int i=1; i<histSize;i++)
    {
        line(histImage, Point(bin_w*(i-1), hist_h-30-cvRound(hist.at<float>(i-1))),
            Point( bin_w*(i), hist_h - 30 - cvRound(hist.at<float>(i))),
            Scalar(0,0,255),2,8,0);
        
        if(i%2==0)
        {
            char buffer[5];
            sprintf(buffer,"%d",i);
            line(histImage, Point(i*bin_w, hist_h - 35), Point(i*bin_w, hist_h - 25), Scalar(0, 0, 0), 2, 8, 0);
            putText(histImage, buffer, Point(i*bin_w, hist_h-5), FONT_HERSHEY_COMPLEX, .5, Scalar(0,0,0), 1, LINE_AA);
        }
    }


    imshow("Histogram", histImage);

    waitKey(0);
    destroyAllWindows();

    return EXIT_SUCCESS;
}

File: classes02/python/hueHistogram.py
import cv2
import argparse
import numpy as np

img=cv2.imread("../data/images/capsicum.jpg")
hsvImage=cv2.cvtColor(img,cv2.COLOR_BGR2HSV)

hsvImage=np.float32(hsvImage)

H,S,V=cv2.split(hsvImage)

plot_width = 540
plot_height = 400
actualRange = 180
rangeRatio = int(plot_width/actualRange)
bufferHeight = 30

histImage = 255*np.ones((plot_height + bufferHeight, plot_width, 3))
xAxisValues = np.arange(plot_width)

histogram = cv2.calcHist([H*rangeRatio],[0],None,[plot_width],[0,plot_width])

cv2.normalize(histogram, histogram, 0, plot_height, cv2.NORM_MINMAX, -1)

histogram = np.max(histogram)-histogram

points=np.column_stack((xAxisValues, histogram.T[0]))

cv2.polylines(histImage, np.int32([points]), False, (0,0,255), 2, cv2.LINE_AA)

font = cv2.FONT_HERSHEY_SIMPLEX
fontScale = 0.4
fontWidth = 1
xInterval = 20

cv2.putText(histImage, "0", (0, int(histImage.shape[0] - (bufferHeight/2)) ), font, fontScale, (0, 0, 0), fontWidth)

for i in range(0,plot_width,rangeRatio*xInterval):
  
  # Specify the position ( xval, yval )for putting the point on x axis 
  xval = i - 7  
  yval = int(histImage.shape[0] - (bufferHeight/2))
  cv2.putText(histImage, str( i/rangeRatio ), (xval, yval ), font, fontScale, (0, 0, 0), fontWidth)

cv2.namedWindow("Original Image", cv2.WINDOW_AUTOSIZE)
cv2.namedWindow("Histogram of Hue channel", cv2.WINDOW_AUTOSIZE)
cv2.imshow("Original Image",img)
cv2.imshow("Histogram of Hue channel",histImage)
cv2.waitKey(0)
cv2.destroyAllWindows()

